addons/mail/models/mail_followers.py

```diff
@@ -42,9 +42,17 @@ class Followers(models.Model):
     @api.multi
     def _invalidate_documents(self):
         """ Invalidate the cache of the documents followed by ``self``. """
-        for record in self:
-            if record.res_id:
-                self.env[record.res_model].invalidate_cache(ids=[record.res_id])
+        to_invalidate = {}
+        # use read for performance:
+        # only prefetch what is needed and avoid calling fields __get__
+        for record in self.read(['res_id', 'res_model']):
+            if record['res_id']:
+                if record['res_model'] not in to_invalidate:
+                    to_invalidate[record['res_model']] = []
+                to_invalidate[record['res_model']].append(record['res_id'])
+        # invalidate in batch for performance
+        for res_model, res_ids in to_invalidate.items():
+            self.env[res_model].invalidate_cache(ids=res_ids)
```

Comment:
I wonder if we shouldn't explicitly invalidate:
message_follower_ids
message_is_follower
message_partner_ids
message_channel_ids
but there is maybe a good reason why it isn't the case

Reply:
Unfortunately everything should be invalidated, as explained just above the method, because access right can change.

I also had trouble to find this comment for some reason... Maybe it should be moved into docstring of the method.

Reply:
Right, I should have read that.